README.md
# FierceCoder
A thin Swift wrapper around the NSCoding protocol.

NOTE: This is still very much a work in progress.

## Features

This wrapper provides a Swift protocol and two classes to encode and decode data. This allows for Swift Structs or Classes to use NSCoding without having to conform directly to that protocol and the Objective-C requirements that entails.

## Requirements

- iOS 7.0+ / Mac OS X 10.9+
- Xcode 7

## Installation

CocoaPods and Carthage support coming soon (Carthage may just work, I haven't tested it yet). Otherwise simply add the FierceCoder project to your workspace and set the FierceCoder as a framework reference.

---

## Usage

### Encoding primitive objects

Any object that is bridged to an Objective-C object that can be encoded by the NSKeyedArchiver can be directly encoded by the FierceCoder. The FierceEncoder provides an encodedData() function that returns an NSData instance containing the encoded data. The FierceDecoder takes in an encoded NSData value as a parameter to the init and allows access to the decoded objects.

```swift
import FierceCoder

let string = "A string"
let number = 42
let goodTime = [8,6,7,5,3,0,9]

let coder = FierceEncoder()
coder.encodeObject(string,key:"a_string")
coder.encodeObject(number,key:"a_number")
coder.encodeObject(goodTime,key:"an_array")
let encodedData = coder.encodedData()

let decoder = FierceDecoder(data: encodedData)
let decodedString = decoder.decodeObject("a_string") as! String
let decodedNumber = decoder.decodeObject("a_number") as! Int
let decodedArray = decoder.decodeObject("an_array") as! [Int]
```

### Encoding Swift Struct or Class

To encode a Swift Struct or Class, just implement the FierceEncodeable protocol. It has a method for encoding and an init that can be overridden to provide decoding. Once this protocol is implemented, the Struct or Class can be encoded by using the .encode and .decode methods on the FierceEncoder and FierceDecoder.

```swift
struct StructTest : FierceEncodeable {
    let name:String
    let title:String
    let age:Int

    private enum EncoderKeys : String {
        case Name
        case Title
        case Age
    }

    init(name:String,title:String,age:Int) {
        self.name = name
        self.title = title
        self.age = age
    }

    init(decoder:FierceDecoder) {
        self.name = decoder.decodeObject(EncoderKeys.Name.rawValue) as! String
        self.title = decoder.decodeObject(EncoderKeys.Title.rawValue) as! String
        if let age = decoder.decodeObject(EncoderKeys.Age.rawValue) as? Int {
            self.age = age
        }
        else {
            self.age = 0
        }
    }

    func encode(encoder: FierceEncoder) {
        encoder.encodeObject(self.name, key: EncoderKeys.Name.rawValue)
        encoder.encodeObject(self.title, key: EncoderKeys.Title.rawValue)
        encoder.encodeObject(self.age, key: EncoderKeys.Age.rawValue)
    }
}
```

### Encoding an array of Structs or Classes

The FierceEncoder and FierceDecoder classes provide methods for encoding and decoding an array of objects that implement the FierceEncodeable protocol. Note that the decoder for an array returns an array of FierceDecoder objects that can be used to construct the decoded objects.

```swift

let myStructs = [StructTest(name: "fred", title: "rock crusher", age: 45),StructTest(name: "barney", title: "rock crusher", age: 40)]

let encoder = FierceEncoder()
encoder.encodeArray(myStructs,key:"array")
let encodedData = coder.encodedData()

let decoder = FierceDecoder(data: encodedData)
let encodedObjects = decoder.decodeArray("array")
var encodedStructs = Array<StructTest>()
for encodedObject in encodedObjects {
    encodedStructs.append(StructTest(decoder: encodedObject))
}
```


## License

FierceCoder is released under the MIT license. See LICENSE for details.


CHANGELOG.md
# FierceCoder CHANGELOG

## 0.0.1

Initial release. Ready for testing but still needs documentation and example application.


FierceCoder/FierceDecoder.swift
//
//  FierceDecoder.swift
//  FierceCoder
//
//  Created by David House on 7/24/15.
//  Copyright © 2015 David House. All rights reserved.
//

import Foundation

public class FierceDecoder {

    var decoder:NSKeyedUnarchiver
    
    public init(data:NSData) {
        self.decoder = NSKeyedUnarchiver(forReadingWithData: data)
    }
    
    public func decode(key:String = "fierceRootObject") -> FierceDecoder {
        if let rawObject = self.decoder.decodeObjectForKey(key) as? NSData {
            let innerDecoder = FierceDecoder(data: rawObject)
            return innerDecoder
        }
        else {
            let innerDecoder = FierceDecoder(data: NSData())
            return innerDecoder
        }
    }

    public func decodeObject(key:String = "fierceRootObject") -> AnyObject? {
        if let object = self.decoder.decodeObjectForKey(key) {
            return object
        }
        else {
            return nil
        }
    }
    
    public func decodeArray(key:String = "fierceRootObject") -> [FierceDecoder] {
        var decoders = Array<FierceDecoder>()
        if let encodedObjects = self.decoder.decodeObjectForKey(key) as? [NSData] {
            for encodedData in encodedObjects {
                decoders.append(FierceDecoder(data: encodedData))
            }
        }
        return decoders
    }
}

FierceCoder/FierceEncoder.swift
//
//  FierceCoder.swift
//  FierceCoder
//
//  Created by David House on 7/24/15.
//  Copyright © 2015 David House. All rights reserved.
//

import Foundation

public class FierceEncoder {
    
    var encoded: NSMutableData?
    var encoder: NSKeyedArchiver
    
    public init() {
        self.encoded = NSMutableData()
        self.encoder = NSKeyedArchiver(forWritingWithMutableData:self.encoded!)
    }
    
    public func encodedData() -> NSData {
        self.encoder.finishEncoding()
        return self.encoded!
    }
    
    public func encode(object:FierceEncodeable,key:String = "fierceRootObject") {
        let objectEncoder = FierceEncoder()
        object.encode(objectEncoder)
        self.encoder.encodeObject(objectEncoder.encodedData(), forKey: key)
    }

    public func encodeObject(object:AnyObject,key:String = "fierceRootObject") {
        self.encoder.encodeObject(object, forKey: key)
    }
    
    public func encodeArray(objects:[FierceEncodeable],key:String = "fierceRootObject") {
        var encodedObjects = Array<NSData>()
        for object in objects {
            let objectEncoder = FierceEncoder()
            object.encode(objectEncoder)
            encodedObjects.append(objectEncoder.encodedData())
        }
        self.encoder.encodeObject(encodedObjects, forKey: key)
    }
}

FierceCoderTests/FierceCoderArrayTests.swift
//
//  FierceCoderArrayTests.swift
//  FierceCoder
//
//  Created by David House on 7/27/15.
//  Copyright © 2015 David House. All rights reserved.
//

import XCTest
@testable import FierceCoder

struct StructWithArrayTest : FierceEncodeable {
    let name:String
    let values:[String]
    
    private enum EncoderKeys : String {
        case Name
        case Values
    }
    
    init(name:String,values:[String]) {
        self.name = name
        self.values = values
    }
    
    init(decoder:FierceDecoder) {
        self.name = decoder.decodeObject(EncoderKeys.Name.rawValue) as! String
        self.values = decoder.decodeObject(EncoderKeys.Values.rawValue) as! [String]
    }
    
    func encode(encoder: FierceEncoder) {
        encoder.encodeObject(self.name, key: EncoderKeys.Name.rawValue)
        encoder.encodeObject(self.values, key: EncoderKeys.Values.rawValue)
    }
}

class FierceCoderArrayTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testEncodeStringArray() {
        
        let strings = ["First","Second","Third","Fourth"]
        let encoder = FierceEncoder()
        encoder.encodeObject(strings)
    
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        if let decodedStrings = decoder.decodeObject() as? [String] {
            XCTAssertTrue(strings == decodedStrings)
        }
        else {
            XCTFail()
        }
    }
    
    func testEncodeStructWithArray() {
        
        let testStruct = StructWithArrayTest(name: "Fred", values: ["Barney","Wilma","Betty"])
        let encoder = FierceEncoder()
        testStruct.encode(encoder)
        
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        let decodedStruct = StructWithArrayTest(decoder: decoder)
        XCTAssertTrue(decodedStruct.name == testStruct.name)
        XCTAssertTrue(decodedStruct.values == testStruct.values)
    }
    
}


FierceCoderTests/FierceCoderChildObjectTests.swift
//
//  FierceCoderChildObjectTests.swift
//  FierceCoder
//
//  Created by David House on 7/27/15.
//  Copyright © 2015 David House. All rights reserved.
//

import XCTest

struct ChildStructTest : FierceEncodeable {

    
    let name:String
    let age:Int
    
    private enum EncoderKeys : String {
        case Name
        case Age
    }

    init(name:String,age:Int) {
        self.name = name
        self.age = age
    }
    
    init(decoder:FierceDecoder) {
        if let name = decoder.decodeObject(EncoderKeys.Name.rawValue) as? String {
            self.name = name
        }
        else {
            self.name = ""
        }
        
        if let age = decoder.decodeObject(EncoderKeys.Age.rawValue) as? Int {
            self.age = age
        }
        else {
            self.age = 0
        }
    }
    
    func encode(encoder: FierceEncoder) {
        encoder.encodeObject(self.name, key: EncoderKeys.Name.rawValue)
        encoder.encodeObject(self.age, key: EncoderKeys.Age.rawValue)
    }
}

struct ParentStructTest : FierceEncodeable {
    let name:String
    let child:ChildStructTest
    
    private enum EncoderKeys : String {
        case Name
        case Child
    }
    
    init(name:String,childName:String,childAge:Int) {
        self.name = name
        self.child = ChildStructTest(name: childName, age: childAge)
    }
    
    init(decoder:FierceDecoder) {
        self.name = decoder.decodeObject(EncoderKeys.Name.rawValue) as! String
        self.child = ChildStructTest(decoder: decoder.decode(EncoderKeys.Child.rawValue))
    }
    
    func encode(encoder: FierceEncoder) {
        encoder.encodeObject(self.name, key: EncoderKeys.Name.rawValue)
        encoder.encode(self.child, key: EncoderKeys.Child.rawValue)
    }
}

struct ParentWithChildArray : FierceEncodeable {
    
    var children = [ChildStructTest]()
    
    init(children:[ChildStructTest]) {
        self.children = children
    }
    
    init(decoder: FierceDecoder) {
        self.children = [ChildStructTest]()
        let encodedChildren = decoder.decodeArray("children")
        for encodedChild in encodedChildren {
            self.children.append(ChildStructTest(decoder: encodedChild))
        }
    }
    
    func encode(encoder: FierceEncoder) {
        encoder.encodeArray(self.children.map{ $0 as FierceEncodeable }, key: "children")
    }
}

class FierceCoderChildObjectTests: XCTestCase {

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testWithChild() {
        
        let parent = ParentStructTest(name: "Fred", childName: "Pebbles", childAge: 10)
        let encoder = FierceEncoder()
        parent.encode(encoder)
        
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        let decodedParent = ParentStructTest(decoder: decoder)
        XCTAssertTrue(parent.name == decodedParent.name)
    }
    
    func testParentWithChildArray() {
        
        let children = [ChildStructTest(name: "fred", age: 12),ChildStructTest(name: "barney", age: 21)]
        let parent = ParentWithChildArray(children: children)
        let encoder = FierceEncoder()
        parent.encode(encoder)
        
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)

        let decoder = FierceDecoder(data: rawData)
        let decodedParent = ParentWithChildArray(decoder: decoder)
        XCTAssertTrue(children.count == decodedParent.children.count)
    }
}


FierceCoderTests/FierceCoderTests.swift
//
//  FierceCoderTests.swift
//  FierceCoderTests
//
//  Created by David House on 7/24/15.
//  Copyright © 2015 David House. All rights reserved.
//

import XCTest
@testable import FierceCoder

struct StructTest : FierceEncodeable {
    let name:String
    let title:String
    let age:Int
    
    private enum EncoderKeys : String {
        case Name
        case Title
        case Age
    }
    
    init(name:String,title:String,age:Int) {
        self.name = name
        self.title = title
        self.age = age
    }
    
    init(decoder:FierceDecoder) {
        self.name = decoder.decodeObject(EncoderKeys.Name.rawValue) as! String
        self.title = decoder.decodeObject(EncoderKeys.Title.rawValue) as! String
        if let age = decoder.decodeObject(EncoderKeys.Age.rawValue) as? Int {
            self.age = age
        }
        else {
            self.age = 0
        }
    }
    
    func encode(encoder: FierceEncoder) {
        encoder.encodeObject(self.name, key: EncoderKeys.Name.rawValue)
        encoder.encodeObject(self.title, key: EncoderKeys.Title.rawValue)
        encoder.encodeObject(self.age, key: EncoderKeys.Age.rawValue)
    }
}

final class ClassTest : FierceEncodeable {
    let name:String
    let title:String
    
    private enum EncoderKeys : String {
        case Name
        case Title
    }
    
    init(name:String,title:String) {
        self.name = name
        self.title = title
    }

    init(decoder:FierceDecoder) {
        self.name = decoder.decodeObject(EncoderKeys.Name.rawValue) as! String
        self.title = decoder.decodeObject(EncoderKeys.Title.rawValue) as! String
    }
    
    func encode(encoder: FierceEncoder) {
        encoder.encodeObject(self.name, key: EncoderKeys.Name.rawValue)
        encoder.encodeObject(self.title, key: EncoderKeys.Title.rawValue)
    }
}

class FierceCoderTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testEncodingObjects() {
        
        let string = "A string"
        let number = 42
        let goodTime = [8,6,7,5,3,0,9]
        
        let coder = FierceEncoder()
        coder.encodeObject(string,key:"a_string")
        coder.encodeObject(number,key:"a_number")
        coder.encodeObject(goodTime,key:"an_array")
        let encodedData = coder.encodedData()
        
        let decoder = FierceDecoder(data: encodedData)
        let decodedString = decoder.decodeObject("a_string") as! String
        let decodedNumber = decoder.decodeObject("a_number") as! Int
        let decodedArray = decoder.decodeObject("an_array") as! [Int]
        XCTAssertTrue(string == decodedString)
        XCTAssertTrue(number == decodedNumber)
        XCTAssertTrue(goodTime.count == decodedArray.count)
        
        let myStructs = [StructTest(name: "fred", title: "rock crusher", age: 45),StructTest(name: "barney", title: "rock crusher", age: 40)]

    }
    
    func testEncodingString() {
        
        let encoder = FierceEncoder()
        let original = "This is the input string"
        encoder.encodeObject(original)

        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        if let after = decoder.decodeObject() as? String {
            XCTAssertTrue(original == after)
        }
        else {
            XCTFail()
        }
    }
    
    func testDifferentKeys() {
        
        let encoder = FierceEncoder()
        let original = "This is the input string"
        encoder.encodeObject(original,key:"original")
        
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        if let after = decoder.decodeObject("after") as? String {
            XCTAssertFalse(original == after)
        }
    }
    
    func testBasicStructEncoding() {
        
        let aStruct = StructTest(name: "Fred", title: "Rock Crusher", age: 45)
        let encoder = FierceEncoder()
        aStruct.encode(encoder)
        
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        let bStruct = StructTest(decoder: decoder)
        XCTAssertTrue(aStruct.name == bStruct.name)
        XCTAssertTrue(aStruct.title == bStruct.title)
        XCTAssertTrue(aStruct.age == bStruct.age)
    }
    
    func testBasicClassEncoding() {
        
        let aStruct = ClassTest(name: "Fred", title: "Rock Crusher")
        let encoder = FierceEncoder()
        aStruct.encode(encoder)
        
        let rawData = encoder.encodedData()
        XCTAssertNotNil(rawData)
        XCTAssertTrue(rawData.length > 0)
        
        let decoder = FierceDecoder(data:rawData)
        let bStruct = ClassTest(decoder: decoder)
        XCTAssertTrue(aStruct.name == bStruct.name)
        XCTAssertTrue(aStruct.title == bStruct.title)
    }
}
